### ISSUE TITLE
Windows 10 Docker processes consuming high CPU with no running containers

### ISSUE THREAD

**Issue by mlmeyers** [docker/for-win#1772]
### Expected behavior
When stopping all containers I would expect CPU usage to drop to normal levels.
### Actual behavior
Almost no change to CPU after stopping containers, further after restarting docker it continues to consume a large percentage of CPU. Consumed by both com.docker.service.exe and Docker for Windows.exe
### Information

A diagnostic was uploaded with id: 4898CA71-860C-4EE7-85B8-E309B53354D8/2018-03-01_13-17-04

Windows Build 16299.248

This is snapshot of my cpu consumption with no running containers. Looking at process explorer it appears the cpu is taken up by threads in clr.dll!DllCanUnloadNowInternal 

Taskmanager 

![image](https://user-images.githubusercontent.com/12535482/36864860-78d3a1d0-1d53-11e8-90d6-1948ab02112d.png)

Docker for windows.exe procexp

![image](https://user-images.githubusercontent.com/12535482/36864955-ca198d34-1d53-11e8-8f4b-b58ed670ddef.png)

com.docker.service.exe  procexp

![image](https://user-images.githubusercontent.com/12535482/36865103-28f38f08-1d54-11e8-9dad-d49b33446455.png)

### Steps to reproduce the behavior

  1. Start docker, start windows containers that do stuff (consume cpu) stop containers.

**Comment by mlmeyers** [comment]
This seems related to the com.docker.service.exe, so far when running just dockerd via cmdline I am unable to reproduce the high cpu issues. In fact when running just the daemon the container hardly puts a dent in my cpu usage. In numerous runs using the docker for windows client the cpu issue is reproduced on my machine. Must say the windows 10 install feels bloated. Wish it was more aligned with Server 2016 setup. The introduction of the com.docker.service and the behind the scenes work to recreate the daemon service etc based on the client tool seems to muddy the waters a lot in comparison to a simple start of the daemon and a config file. Still not clear on the cause for the high cpu but suspect some threading issue trying to unload dll's  still in use based on the active threads in procexp.

**Comment by mlmeyers** [comment]
Possible to get eyes on this I continue to struggle with Docker for Windows service consuming over half of my CPU when running. Always the cycles are being consumed by clr.dll!DllCanUnloadNowInternal

**Comment by guba-odudkin** [comment]
Same issue here. Started after I updated to the latest version. Should I rollback? Otherwise it's not usable

**Comment by SonicGD** [comment]
Same issue with 18.03 and windows skip ahead build. 

![image](https://user-images.githubusercontent.com/1222013/38185604-6e6536b8-3668-11e8-94fa-8c4272d6dfc8.png)

**Comment by smasherprog** [comment]
Same issue here on windows 10 1709

**Comment by cocowalla** [comment]
I've got this on Edge build 18.05 too

**Comment by rapatel0** [comment]
Same Issue here
![image](https://user-images.githubusercontent.com/18704704/40047381-3bf8d528-5827-11e8-896e-a05aa75e6eca.png)

**Comment by OpaDopa** [comment]
Same here

**Comment by ddobric** [comment]
Anything new here?

**Comment by guba-odudkin** [comment]
Stopped using docker

**Comment by ddobric** [comment]
@guba-odudkin this is radical decision. But I can understand,

**Comment by industrial64** [comment]
Same on 19098 - No resolution yet. High use with or without running containers.
Seen this only on Win10.

**Comment by piyh** [comment]
Getting this as well.

**Comment by DealWithMeal** [comment]
Same stuff here. A **fix**?

**Comment by industrial64** [comment]
My fix was to increase memory due to Docker for Win not allocating enough to run kubernetes and a single deb container. Just increase mem (mine from 2.2GB to 2.8GB) reload, and everything should go back to normal CPU-wise.

**Comment by piyh** [comment]
There aren't any containers running, it's only docker.  Like rapatel0's screenshot.

**Comment by industrial64** [comment]
Yup, same fix applies. I had nothing running, increasing memory reserves fixed the CPU issue (w/ and w/o containers). Just give it a whirl.

**Comment by whitecolor** [comment]
Advice: try to restart Docker for Windows Service in services.msc, then restart docker for windows, in Settings/Reset.

**Comment by dannyk81** [comment]
thanks @whitecolor, but I tried restart, reseting, reinstalling, everything... nothing helps 😟 d4w became totally unusable for me 😔

**Comment by whitecolor** [comment]
I see such pattern, as I rarely tend to restart windows sometimes (after windows a certain restart) docker works fine (without consuming CPU and mem) for a long time, but sometimes after a few hours of work or idle it starts consuming CPU/mem, then service restart helps. Also I need to restart docker by unchecking and checking again `Exspose deamon` setting - without it `localhost:2375` is not available after the restart. =) Quite cumbersome, but it still tends to work, though I'm still currently on `Version 17.09.1-ce-win42 (14687)`

**Comment by dannyk81** [comment]
thanks again @whitecolor, but the restart tricks are not working for me 😢 

Windows just got updated to build 18252, however the issue still persists and docker logs now shows constantly repeating:
```
[13:11:53.398][DnsUpdater     ][Info   ] Network configuration change detected
[13:11:53.399][DnsUpdater     ][Error  ] Unable to update dns settings: Value cannot be null.
Parameter name: first
```

This is so frustrating...

/cc @djs55

**Comment by antogh** [comment]
Same problem here , docker service eating 15% cpu for doing nothing , all containers are stopped. My fix is to quit docker and start it again when I need it .

**Comment by mts57017** [comment]
same issue here. Killing docker and using when i need too.

**Comment by FrankSzendzielarz** [comment]
Likewise started just this evening. Nothing changed AFAICT

**Comment by ma1f** [comment]
uninstall/reinstall with linux containers instead of windows containers - then once running switch to windows containers.
Fixed the issue for me.

**Comment by dannyk81** [comment]
@ma1f good suggestions, for some reason I didn't consider LCOW as an option, this indeed solved the CPU option and seems to be good for my use cases, not sure what limitations there might be, but for general purpose linux development stuff looks ok so far.

**Comment by ma1f** [comment]
@dannyk81 it actually seems to set it up better if anything when installed as linux containers - eg. the hyper-v virtual network properly creates the DockerNAT, when installing as windows it doesn't...
Once running i've switched to windows containers from the menu and had no issues / no high CPU usage and seems to work great.

**Comment by DealWithMeal** [comment]
Reinstalling with linux containers and switching to windows containers didn't solve the issue for me :(

**Comment by dannyk81** [comment]
@ma1f this looked like a promising solution and in my case the high CPU usage issue is gone after switching to LCOW, but now I'm hitting another issue 😞 

Looks like DNS resolution is broken... I have a dns resolver running on my host (192.168.2.13), it is picked up correctly in /etc/resolv.conf of the container:
```
/ # cat /etc/resolv.conf
nameserver 172.29.144.1
nameserver 192.168.2.13
```

But for some reason I just can't access it from the containers.

from Windows:
```
C:\WINDOWS\system32>nslookup google.com
Server:  UnKnown
Address:  192.168.2.13

Non-authoritative answer:
Name:    google.com
Address:  172.217.9.46
```

From a container:
```
/ # dig @192.168.2.13 google.com

; <<>> DiG 9.10.2 <<>> @192.168.2.13 google.com
; (1 server found)
;; global options: +cmd
;; connection timed out; no servers could be reached
```

I can query external dns servers (like 8.8.8.8 or 1.1.1.1) but the one running on the host itself... this is so frustrating...

**Comment by cocowalla** [comment]
@ma1f @dannyk81 a good reason LCOW hasn't worked for me at least, is that LCOW has some serious limitations that mean a lot of Linux containers don't work. Specifically, `chmod`, `chown` and `chattr` don't work, and these fairly fundamental tools are essential for a lot of containers. See [here](https://github.com/moby/moby/issues/35665) for more info.

**Comment by dannyk81** [comment]
@cocowalla good points!

well... I'm out of ideas, short of installing Linux 😉

Considering the amount of people reporting of this issue here and on https://github.com/docker/for-win/issues/1900, I would expect someone to take a look at this...

**Comment by ma1f** [comment]
@cocowalla interesting limitations on LCOW, haven't had issues yet in the way i've been using them but useful to know, i'm using both linux and windows but key thing is that reinstalling without ticking the *use Windows Containers* resolves the CPU issue.
@DealWithMeal for me I only realised this was the issue as none of my work colleagues were having issues but none of them had selected the *use Windows Containers* checkbox during installation - perhaps make sure you have done a full uninstall - https://success.docker.com/article/how-to-completely-remove-docker-in-windows-10

**Comment by Dannymx** [comment]
Just came here to report the same thing, high cpu usage, I tried all the suggestions mentioned in this thread but nothing fixed it. Running Version 18.06.1-ce-win73 (19507) Channel: stable.

**Comment by dannyk81** [comment]
for anyone interested, I "solved" my dns issue by adding an additional IP to the `nat` network interface and configured my local resolver to listen on that address.

I had to than configure this IP as the DNS server on all network interfaces so that when a container is spun-up it would be populated into its `/etc/resolv.conf` file.


On a personal note, Windows is becoming a PITA... having to deal with all these issues and obscure configuration stuff...

**Comment by clarity99** [comment]
uninstalled/reinstall without checkbox for windows containers, didn't help. still using 15% cpu constantly

**Comment by mikeparker** [comment]
I'm struggling to reproduce this. If anyone could get specific reproduction instructions that'd be great, including which container(s) you're running and what settings you have for your VM (ram, CPUs etc).

The original post said that it only occurs when running containers then stopping them (it doesn't occur purely on boot), if people could confirm that too that'd be great.

**Comment by mikeparker** [comment]
@dannyk81 Is it possible you can completely disable your VPNs and try? For me the above instructions with no VPN active, all docker processes are at 0% CPU so I can't reproduce.

If it goes to 0 completely for you, it sounds networking related and I can look into reproducing with VPNs, although to reproduce I'm guessing I may need to go through many iterations of trying to set up a VPN to try to get a similar setup so it'd be great to confirm it's definitely related to VPNs before i go down that path!

**Comment by DealWithMeal** [comment]
10-20% ? I mean.. I wouldn't even think that something is wrong in that case.

I have 8th gen I7 CPU and docker starts to consume **70-90%** of it once I start it and do whatever! For example, start and stop any kind of container. In both windows and linux containers, using kubernetes and not.

**Comment by mts57017** [comment]
it is 10-20% with no containers running at all. Basically a fresh install just hanging around. That should not cause a constant usage spike that high, considering it didn't use too. If it does, i would like to know why.

**Comment by dannyk81** [comment]
@mikeparker thanks for looking into this, sorry for the delay, I'm away for few days but can run the test once I get back.

From past experience, if Docker started after a clean boot and no VPN tunnels established, the CPU usage of the docker service was negligible/close to zero.

However, I would like to make a test once I'm back to verify this, will keep you posted.

**Comment by ma1f** [comment]
@mikeparker one thing I did notice is that if installed with windows containers checkbox selected, no `nat` network for hyper-v was created, whereas if installed with linux containers (default) then it is created fine.

**Comment by clarity99** [comment]
Hi, I don't have any VPN tunnels, yet I am experiencing this. It does not matter if I install with linux or windows containers at the beginning. 
I've managed to get the callstack of the thread using the CPU:
`ntdll.dll!NtOpenKeyEx+0x14
KERNELBASE.dll!LocalBaseRegOpenKey+0x1ce
KERNELBASE.dll!RegOpenKeyExInternalA+0x12e
KERNELBASE.dll!RegOpenKeyExA+0x19
IPHLPAPI.DLL!AddDnsServerAddressesToAdapter+0x276
IPHLPAPI.DLL!AddDnsConfiguration+0x4ec
IPHLPAPI.DLL!AllocateAndGetAdaptersAddresses+0x82b
IPHLPAPI.DLL!GetAdaptersAddresses+0x4f
DNSAPI.dll!IpHelp_GetAdaptersAddresses+0xb3
DNSAPI.dll!NetInfo_Build+0xe5
DNSAPI.dll!NetInfo_Get+0x2e0
DNSAPI.dll!Config_GetDnsServerList+0x52
DNSAPI.dll!Config_GetDnsServerListIp4+0x3b
DNSAPI.dll!DnsQueryConfig+0x229f8
IPHLPAPI.DLL!AllocateAndGetPerAdapterInfo+0xf8
IPHLPAPI.DLL!GetPerAdapterInfo+0x30
[Native Frame: IL Method without Metadata]
[Managed to Unmanaged Transition]
System.dll!System.Net.NetworkInformation.SystemIPv4InterfaceProperties.GetPerAdapterInfo+0x66
System.dll!System.Net.NetworkInformation.SystemIPInterfaceProperties..ctor+0x207
System.dll!System.Net.NetworkInformation.SystemNetworkInterface..ctor+0xc7
System.dll!System.Net.NetworkInformation.SystemNetworkInterface.GetNetworkInterfaces+0x191
System.dll!System.Net.NetworkInformation.NetworkInterface.GetAllNetworkInterfaces+0x46
Docker.Backend.dll!Docker.Backend.NetworkInterfaceHelper.GetHostInterfaceByIp+0x3f
Docker.Backend.dll!Docker.Backend.NetworkInterfaceHelper.TryEnableSambaShareOnInterface+0x27
[Unmanaged to Managed Transition]
clr.dll!CallDescrWorkerInternal+0x83
clr.dll!CallDescrWorkerWithHandler+0x4e
clr.dll!CallDescrWorkerReflectionWrapper+0x1a
clr.dll!RuntimeMethodHandle::InvokeMethod+0x45f
[Managed to Unmanaged Transition]
mscorlib.dll!System.Reflection.RuntimeMethodInfo.UnsafeInvokeInternal+0x80
mscorlib.dll!System.Delegate.DynamicInvokeImpl+0x99
System.Management.dll!System.Management.WmiDelegateInvoker.FireEventToDelegates+0x97
System.Management.dll!System.Management.ManagementEventWatcher.FireEventArrived+0x28
System.Management.dll!System.Management.SinkForEventQuery.Indicate+0x8d
[Unmanaged to Managed Transition]
[Native Frame: IL Method without Metadata]
clr.dll!COMToCLRDispatchHelper+0x39
clr.dll!COMToCLRWorker+0x1b4
clr.dll!GenericComCallStub+0x57
wminet_utils.dll!CWmiSinkDemultiplexor::InternalWbemObjectSink::Indicate+0x56
fastprox.dll!CSinkStubBuffer::XSinkStublet::Indicate_Stub+0xf7
combase.dll!ObjectMethodExceptionHandlingAction<<lambda_76d9e92c799d246a4afbe64a2bf5673d> >+0x53
combase.dll!DefaultStubInvoke+0x295
combase.dll!ServerCall::ContextInvoke+0x496
combase.dll!AppInvoke+0x328
combase.dll!ComInvokeWithLockAndIPID+0x59c
combase.dll!ThreadInvoke+0x1320
RPCRT4.dll!DispatchToStubInCNoAvrf+0x24
RPCRT4.dll!RPC_INTERFACE::DispatchToStubWorker+0x1d8
RPCRT4.dll!RPC_INTERFACE::DispatchToStubWithObject+0x154
RPCRT4.dll!LRPC_SCALL::DispatchRequest+0x18d
RPCRT4.dll!LRPC_SCALL::HandleRequest+0x86c
RPCRT4.dll!LRPC_ADDRESS::HandleRequest+0x33d
RPCRT4.dll!LRPC_ADDRESS::ProcessIO+0x8ad
RPCRT4.dll!LrpcIoComplete+0xd8
ntdll.dll!TppAlpcpExecuteCallback+0x22e
ntdll.dll!TppWorkerThread+0x258
KERNEL32.dll!BaseThreadInitThunk+0x14
ntdll.dll!RtlUserThreadStart+0x21
`

so it seems it does have something to do with network interfaces.
ipconfig gives the following output on my system:
`
PS C:\users\roberti> ipconfig /all

Windows IP Configuration

   Host Name . . . . . . . . . . . . : WINDHORSE
   Primary Dns Suffix  . . . . . . . : CREA.SI
   Node Type . . . . . . . . . . . . : Hybrid
   IP Routing Enabled. . . . . . . . : No
   WINS Proxy Enabled. . . . . . . . : No
   DNS Suffix Search List. . . . . . : CREA.SI

Ethernet adapter Local Area Connection 4 2:

   Connection-specific DNS Suffix  . :
   Description . . . . . . . . . . . : SonicWALL Virtual NIC
   Physical Address. . . . . . . . . : 00-60-73-E0-00-00
   DHCP Enabled. . . . . . . . . . . : Yes
   Autoconfiguration Enabled . . . . : Yes
   Link-local IPv6 Address . . . . . : fe80::58d7:9c92:3167:f3e9%2(Preferred)
   Default Gateway . . . . . . . . . :
   DHCPv6 IAID . . . . . . . . . . . : 956326003
   DHCPv6 Client DUID. . . . . . . . : 00-01-00-01-1C-27-32-D2-D0-50-99-35-4D-23
   DNS Servers . . . . . . . . . . . : fec0:0:0:ffff::1%1
                                       fec0:0:0:ffff::2%1
                                       fec0:0:0:ffff::3%1
   NetBIOS over Tcpip. . . . . . . . : Enabled

Ethernet adapter Ethernet 2:

   Media State . . . . . . . . . . . : Media disconnected
   Connection-specific DNS Suffix  . :
   Description . . . . . . . . . . . : Fortinet SSL VPN Virtual Ethernet Adapter
   Physical Address. . . . . . . . . : 00-09-0F-AA-00-01
   DHCP Enabled. . . . . . . . . . . : Yes
   Autoconfiguration Enabled . . . . : Yes

Ethernet adapter vEthernet (DockerNAT):

   Connection-specific DNS Suffix  . :
   Description . . . . . . . . . . . : Hyper-V Virtual Ethernet Adapter #2
   Physical Address. . . . . . . . . : 00-15-5D-00-8D-04
   DHCP Enabled. . . . . . . . . . . : No
   Autoconfiguration Enabled . . . . : Yes
   IPv4 Address. . . . . . . . . . . : 10.0.75.1(Preferred)
   Subnet Mask . . . . . . . . . . . : 255.255.255.0
   Default Gateway . . . . . . . . . :
   DNS Servers . . . . . . . . . . . : fec0:0:0:ffff::1%1
                                       fec0:0:0:ffff::2%1
                                       fec0:0:0:ffff::3%1
   NetBIOS over Tcpip. . . . . . . . : Enabled

Ethernet adapter Local Area Connection:

   Connection-specific DNS Suffix  . : CREA.SI
   Description . . . . . . . . . . . : Intel(R) Ethernet Connection (2) I218-V
   Physical Address. . . . . . . . . : D0-50-99-35-4D-23
   DHCP Enabled. . . . . . . . . . . : Yes
   Autoconfiguration Enabled . . . . : Yes
   Link-local IPv6 Address . . . . . : fe80::8902:f47f:611e:e96e%27(Preferred)
   IPv4 Address. . . . . . . . . . . : 192.168.0.141(Preferred)
   Subnet Mask . . . . . . . . . . . : 255.255.248.0
   Lease Obtained. . . . . . . . . . : 2. november 2018 10:21:45
   Lease Expires . . . . . . . . . . : 9. november 2018 5:33:56
   Default Gateway . . . . . . . . . : 192.168.0.1
   DHCP Server . . . . . . . . . . . : 192.168.0.64
   DHCPv6 IAID . . . . . . . . . . . : 248533145
   DHCPv6 Client DUID. . . . . . . . : 00-01-00-01-1C-27-32-D2-D0-50-99-35-4D-23
   DNS Servers . . . . . . . . . . . : 192.168.0.61
                                       192.168.0.62
   Primary WINS Server . . . . . . . : 192.168.0.62
   Secondary WINS Server . . . . . . : 192.168.0.61
   NetBIOS over Tcpip. . . . . . . . : Enabled

Ethernet adapter VirtualBox Host-Only Network:

   Connection-specific DNS Suffix  . :
   Description . . . . . . . . . . . : VirtualBox Host-Only Ethernet Adapter
   Physical Address. . . . . . . . . : 0A-00-27-00-00-04
   DHCP Enabled. . . . . . . . . . . : No
   Autoconfiguration Enabled . . . . : Yes
   Link-local IPv6 Address . . . . . : fe80::e437:95f2:11a1:1f4b%4(Preferred)
   IPv4 Address. . . . . . . . . . . : 192.168.56.1(Preferred)
   Subnet Mask . . . . . . . . . . . : 255.255.255.0
   Default Gateway . . . . . . . . . :
   DNS Servers . . . . . . . . . . . : fec0:0:0:ffff::1%1
                                       fec0:0:0:ffff::2%1
                                       fec0:0:0:ffff::3%1
   NetBIOS over Tcpip. . . . . . . . : Enabled

Ethernet adapter VirtualBox Host-Only Network #2:

   Connection-specific DNS Suffix  . :
   Description . . . . . . . . . . . : VirtualBox Host-Only Ethernet Adapter #2
   Physical Address. . . . . . . . . : 0A-00-27-00-00-1E
   DHCP Enabled. . . . . . . . . . . : No
   Autoconfiguration Enabled . . . . : Yes
   Link-local IPv6 Address . . . . . : fe80::b8d0:3a50:c3d:38bc%30(Preferred)
   IPv4 Address. . . . . . . . . . . : 192.168.99.1(Preferred)
   Subnet Mask . . . . . . . . . . . : 255.255.255.0
   Default Gateway . . . . . . . . . :
   DNS Servers . . . . . . . . . . . : fec0:0:0:ffff::1%1
                                       fec0:0:0:ffff::2%1
                                       fec0:0:0:ffff::3%1
   NetBIOS over Tcpip. . . . . . . . : Enabled

Ethernet adapter VMware Network Adapter VMnet1:

   Connection-specific DNS Suffix  . :
   Description . . . . . . . . . . . : VMware Virtual Ethernet Adapter for VMnet1
   Physical Address. . . . . . . . . : 00-50-56-C0-00-01
   DHCP Enabled. . . . . . . . . . . : No
   Autoconfiguration Enabled . . . . : Yes
   Link-local IPv6 Address . . . . . : fe80::99a2:f529:ffeb:d6fe%22(Preferred)
   IPv4 Address. . . . . . . . . . . : 192.168.241.1(Preferred)
   Subnet Mask . . . . . . . . . . . : 255.255.255.0
   Default Gateway . . . . . . . . . :
   DNS Servers . . . . . . . . . . . : fec0:0:0:ffff::1%1
                                       fec0:0:0:ffff::2%1
                                       fec0:0:0:ffff::3%1
   NetBIOS over Tcpip. . . . . . . . : Enabled

Ethernet adapter VMware Network Adapter VMnet8:

   Connection-specific DNS Suffix  . :
   Description . . . . . . . . . . . : VMware Virtual Ethernet Adapter for VMnet8
   Physical Address. . . . . . . . . : 00-50-56-C0-00-08
   DHCP Enabled. . . . . . . . . . . : No
   Autoconfiguration Enabled . . . . : Yes
   Link-local IPv6 Address . . . . . : fe80::bc15:19de:a31a:aa1c%12(Preferred)
   IPv4 Address. . . . . . . . . . . : 192.168.137.1(Preferred)
   Subnet Mask . . . . . . . . . . . : 255.255.255.0
   Default Gateway . . . . . . . . . :
   DNS Servers . . . . . . . . . . . : fec0:0:0:ffff::1%1
                                       fec0:0:0:ffff::2%1
                                       fec0:0:0:ffff::3%1
   NetBIOS over Tcpip. . . . . . . . : Enabled

Ethernet adapter Local Area Connection 3:

   Media State . . . . . . . . . . . : Media disconnected
   Connection-specific DNS Suffix  . :
   Description . . . . . . . . . . . : TAP-Windows Adapter V9
   Physical Address. . . . . . . . . : 00-FF-C5-CC-F0-1D
   DHCP Enabled. . . . . . . . . . . : Yes
   Autoconfiguration Enabled . . . . : Yes

Ethernet adapter Ethernet:

   Media State . . . . . . . . . . . : Media disconnected
   Connection-specific DNS Suffix  . :
   Description . . . . . . . . . . . : TeamViewer VPN Adapter
   Physical Address. . . . . . . . . : 00-FF-66-F3-7B-CB
   DHCP Enabled. . . . . . . . . . . : Yes
   Autoconfiguration Enabled . . . . : Yes

Ethernet adapter Ethernet 3:

   Media State . . . . . . . . . . . : Media disconnected
   Connection-specific DNS Suffix  . :
   Description . . . . . . . . . . . : Fortinet virtual adapter #2
   Physical Address. . . . . . . . . : 00-09-0F-FE-00-01
   DHCP Enabled. . . . . . . . . . . : Yes
   Autoconfiguration Enabled . . . . : Yes

Ethernet adapter vEthernet (Default Switch):

   Connection-specific DNS Suffix  . :
   Description . . . . . . . . . . . : Hyper-V Virtual Ethernet Adapter
   Physical Address. . . . . . . . . : F6-15-1E-2C-6D-98
   DHCP Enabled. . . . . . . . . . . : Yes
   Autoconfiguration Enabled . . . . : Yes
   Link-local IPv6 Address . . . . . : fe80::c150:dd41:c19:686c%7(Preferred)
   IPv4 Address. . . . . . . . . . . : 172.23.152.129(Preferred)
   Subnet Mask . . . . . . . . . . . : 255.255.255.240
   Default Gateway . . . . . . . . . :
   DHCPv6 IAID . . . . . . . . . . . : 1644172637
   DHCPv6 Client DUID. . . . . . . . : 00-01-00-01-1C-27-32-D2-D0-50-99-35-4D-23
   DNS Servers . . . . . . . . . . . : fec0:0:0:ffff::1%1
                                       fec0:0:0:ffff::2%1
                                       fec0:0:0:ffff::3%1
   NetBIOS over Tcpip. . . . . . . . : Disabled

Ethernet adapter vEthernet (nat):

   Connection-specific DNS Suffix  . :
   Description . . . . . . . . . . . : Hyper-V Virtual Ethernet Adapter #3
   Physical Address. . . . . . . . . : 00-15-5D-84-F0-73
   DHCP Enabled. . . . . . . . . . . : Yes
   Autoconfiguration Enabled . . . . : Yes
   Link-local IPv6 Address . . . . . : fe80::9491:e6fe:775b:55d2%18(Preferred)
   IPv4 Address. . . . . . . . . . . : 172.26.32.1(Preferred)
   Subnet Mask . . . . . . . . . . . : 255.255.240.0
   Default Gateway . . . . . . . . . :
   DHCPv6 IAID . . . . . . . . . . . : 1946162525
   DHCPv6 Client DUID. . . . . . . . : 00-01-00-01-1C-27-32-D2-D0-50-99-35-4D-23
   DNS Servers . . . . . . . . . . . : fec0:0:0:ffff::1%1
                                       fec0:0:0:ffff::2%1
                                       fec0:0:0:ffff::3%1
   NetBIOS over Tcpip. . . . . . . . : Enabled
PS C:\users\roberti>`

hope this helps. let me know if I can give some more information

**Comment by dannyk81** [comment]
@mikeparker again, sorry for the delay.

I did a fresh install for D4W and restarted Windows, VPNs are not connected and I still see ~8-10% cpu usage on the Docker service.

I then connected to my VPNs and noticed that CPU usage increased slightly to ~10-15% (also the messages I mentioned here: https://github.com/docker/for-win/issues/1772#issuecomment-421586273 started flooding the log file), I'm not sure if there's any clear correlation here but looks like it does make the situation worse, though the issue is still there when VPNs are not connected.

This seems to correlate with reports from many other users here on this thread and over at https://github.com/docker/for-win/issues/1900

Switched back to Windows Containers and CPU usage is now under control, for now this is an acceptable workaround for me, although I do hit the limitations of LCOW from time to time, which is a bit frustrating.

**Comment by jedidja** [comment]
Glad to hear there's some progress on this :) I've experienced the same problem as @mts57017 on two machines.

**Comment by mikeparker** [comment]
Can I get confirmation about your DNS Server setting - are you using `Automatic` or `Fixed` DNS?

**Comment by dannyk81** [comment]
@mikeparker I tried both ways by the way, but didn't see any difference.

**Comment by FrankSzendzielarz** [comment]
Automatic. Though recently the issue stopped and has not resurfaced. I will be keeping an eye on it.

**Comment by KamilSzczepaniuk** [comment]
I just installed newest version 2.0.0.0-win78 (28905), problem with high CPU usage seems to be resolved.

**Comment by jedidja** [comment]
Wow! What a huge difference :) Upgrading to 2.0.0.0-win78 (28905) also fixed the problem for me.

**Comment by aloneguid** [comment]
version 2.0.0.0-win78 (28905) resolved the issue for a day, however after reboot i'm back to the same insane cpu usage with no containers running

**Comment by mikeparker** [comment]
We've identified 2 causes, one for `Fixed DNS` which has been resolved and will be shipped with the next release, I look forward to seeing if this fixes CPU usage for fixed DNS. 

The 2nd fix (for automatic DNS) is more complex and requires us to rewrite a decent portion of our DNS code. This will take longer (1-2 weeks) but is in progress and hopefully I can provide further updates soon.

**Comment by mpnewcomb** [comment]
Any updates?

**Comment by mickdaly** [comment]
Docker on my Windows 10 laptop is unusable.

1. When I had no containers and after pulling containers the `vmmem` process is constantly using 44%+ of the CPU
1. When I tried pulling containers, the process was extremely slow, taking 20 minutes until I cancelled for something I would have expected to take a few minutes. 

Is `vmmem` using CPU related to this issue or should I open another?

**Comment by FrankSzendzielarz** [comment]
Interesting, the latest update I got a few days ago brought this issue back.

**Comment by mikeparker** [comment]
**Update**: We have finished the fixes for both Automatic DNS and Fixed DNS, and these will ship with the next release. This fixes high CPU usage inside networking/DNS related code.

`vmmem` process CPU usage is likely to be a separate issue, so it'd be great if you could open a new issue for that and elaborate on your setup. In particular how much memory you have and how much is allocated to docker, but also whether you are running linux or windows containers, which images you were pulling, which version of Windows 10 you have.

**Comment by mikeparker** [comment]
The DNS fixes were released on Edge channel last week. It'd be great if anyone could confirm if this has made any difference to CPU usage when Docker Desktop on Windows is idle, in particular anyone seeing VPN related issues, @dannyk81 @clarity99

**Comment by XzenTorXz** [comment]
it fixed the issue for me!

**Comment by mikeparker** [comment]
@XzenTorXz @dannyk81 great news, thanks for the update!

@tauinger-de that sounds like a different problem, can you open a new ticket please?

**Comment by tauinger-de** [comment]
@mikeparker Yes, can do, however please note that the CPU load remains high without **any** container running..! Which is the title of this ticket..

**Comment by mikeparker** [comment]
@tauinger-de I agree this ticket is a bit vague and there's potentially multiple issues at play, but it's much easier to track if there are separate tickets for each problem. It also means people which issues are now fixed won't get spammed with any further thread replies. This issue is already up to 67 replies and we've found and fixed 2 bugs already.

Also it should give you the template to report the full diagnostic and reproduction instructions so we don't need to go back and forth asking for all your details.

Thanks!

**Comment by Happyman0815** [comment]
I just closed the "Docker Deamon" via the Docker Settings GUI (un-check "Expose Docker Deamon on TCP ... without TLS") and then closed Docker Desktop.  vmmem.exe stopped consuming CPU after this.

**Comment by ArneMancofi** [comment]
@Happyman0815, I had the same experience. In my case, I had Rider running with the Docker Integration plugin connected to the Docker server via port 2375. This resulted in 2.6 HTTP GET requests per second, transferring 40Mbyte per minute over the loopback connection to port 2375.
Naturally, disconnecting the Rider plugin from the Docker server also reduced the CPU load.

**Comment by bobhy** [comment]
I tried the above workarounds: check option to expose daemon without TLS; increase hyper v memory allocated to Moby VM (leaving it statically allocated); enable fixed DNS (8.8.8.8).  I was careful to reboot Windows PC after each one.  None of these helped me, symptom of 15% CPU consumed by Moby VM persisted.  

Last thing: I switched from Linux to Windows containers and rebooted.  This make the excess CPU disappear.  Of course Moby VM is not running, but no other docker-related process is consuming excess CPU now.

**Comment by bobhy** [comment]
Confirming, this problem is resolved in edge release 2.0.3.0 (31778), with default settings for TLS, VM memory and automatic DNS.

**Comment by whitecolor** [comment]
And I can confirm that 2.0.3.0 (31778), STILL HAS the same issue with taking more and more memory and consuming CPU.

**Comment by ardalis** [comment]
2.0.3.0 (3177) here. com.docker.service memory allocation is out of control,resulting in constant page faults and pagefile usage, bringing the whole machine to a craw. Help?
![image](https://user-images.githubusercontent.com/782127/58272488-a3cb8c80-7d5c-11e9-9aa6-db412f443712.png)

**Comment by AaronFriel** [comment]
@mikeparker I've created a separate ticket (#3858) and I have not seen a reply from anyone at Docker. Would appreciate if someone could take a look and ack that it's being tracked internally! This bug keeps my work laptop from ever entering an idle state, which results in extremely short battery life with no apps running.

**Comment by benyaminl** [comment]
> @mikeparker I've created a separate ticket (#3858) and I have not seen a reply from anyone at Docker. Would appreciate if someone could take a look and ack that it's being tracked internally! This bug keeps my work laptop from ever entering an idle state, which results in extremely short battery life with no apps running.

I also exprience it for no reason. Ugh.. Feels like Docker need to work mor hard on windows than on Linux.

**Comment by mikeparker** [comment]
Hi all. We found and fixed a few different bugs in this area and users that had issues with those bugs are reporting they are fixed, so I'm closing this ticket.

I appreciate there may be _other_ issues around CPU and memory usage, but we need to track these in separate tickets. In particular we know we need to:

a) Improve our diagnostics for users to view resource usage of running containers
b) Work with Kubernetes to ensure it runs on windows without high CPU usage

If anyone has further issues please open a new ticket and include diagnostics.

**Comment by FrankSzendzielarz** [comment]
This is back.

**Comment by jwhipp** [comment]
Same issue here.   No containers, MobyLinux VM is chewing 1 CPU core.   I can mitigate the issue by issuing the restart command through the docker systray icon.

**Comment by cocowalla** [comment]
With the announcement today about the future of Docker Desktop on WSL2, I think it's safe to say this isn't going to be fixed.

On the bright side, once WSL2 and the new Docker Desktop lands, we should finally have a decent Docker experience on Windows :smile:

**Comment by elertan** [comment]
Also experiencing this on MacOS.

Docker Desktop Community: 2.0.0.3 (31259)
Channel: stable
8858db33c8
Engine: 18.09.2

Seems to be less platform related now...

> 
> 
> Same issue here. No containers, MobyLinux VM is chewing 1 CPU core. I can mitigate the issue by issuing the restart command through the docker systray icon.
> 
> Also, k8s is not enabled on my install.

**Comment by ziilok** [comment]
Windows Server 2019. Docker 18.09.7 - ~90% cpu usage. Docker daemon mostly used for building images (isolation = process).

**Comment by cristianmad** [comment]
Increasing the memory to ~3GB fixed it for me. Running v2.1.0.1 (37199). I should mention that I did not have this problem until a few days ago when I upgraded to the latest version (the one mentioned above).

**Comment by whitecolor** [comment]
> Increasing the memory to ~3GB fixed it for me

Increasing the resources probably may just delay the consequences.

**Comment by cristianmad** [comment]
Well, it 'fixed' it for now. :) I'll take any other recommendations. I tried other things and none of them worked.

**Comment by meelos** [comment]
Same problem here. Win 10 Build 18970, no containers and using 1,300 MB Memory.

**Comment by StingyJack** [comment]
@mikeparker - please reopen this. Anything we report isn't going to carry a better or different description, and an issue's longevity is an important factor in prioritization. 

Also, new docker related issues tend to fall victim to the evil GitHub docker bot.

**Comment by mikeparker** [comment]
@StingyJack 
> Anything we report isn't going to carry a better or different description
Actually, this isn't true. We've had 11 comments since i closed this issue and none of the comments have included a diagnostic ID, which is 100% necessary in order for us to take any action. New issues pretty much always contain diagnostic IDs (or they get closed / ignored), so opening new issues will very likely give us more information.

Re-opening this isn't really helpful or accurate. We have fixed the original issue which was a problem with the DNS code refreshing too frequently based on network adapter events. Re-opening an old fixed issue is not the way tickets get prioritised at Docker.

> an issue's longevity is an important factor in prioritization.

Actually the opposite is true. New bugs are more likely to be worked on as they are likely recent regressions and something we can easily fix. Furthermore, I am part of prioritization at docker and I can assure you, adding 'me too' comments into an old closed bug will not help us fix any issue, especially if we can't reproduce it and we can't see what setup you're running with.

I know it's hard to believe when your local setup is not working, but the vast majority of users do not see high CPU usage problems, so problems like this are almost always related to a local setup problem, typically one of: hardware misbehaving, using Docker Desktop in nested virtualized environments or 3rd party programs interfering with docker (VPNs, Firewalls, antivirus..). Thus I need to see detailed diagnostics as to what exactly your setup is and look at each case individually in detail. Whats your hardware, which features are you using, what containers are you running, are you running on a server or a desktop, what do your logs say, are there errors there, how fast is the machine running...

Please understand, I do actually want to help. Please open a new ticket and attach a diagnostic.

**Comment by cristianmad** [comment]
@mikeparker 
Thank you Mike,

What kind of logs/diagnostics would be needed? My system is experiencing this right now. I will upload anything that might help.
Thanks
![image](https://user-images.githubusercontent.com/5718092/64181432-105e2900-ce2c-11e9-84fc-f15d3cb84aff.png)

**Comment by mikeparker** [comment]
Docker Desktop ships with a detailed diagnostic gathering tool which will upload everything we need.
Open docker desktop -> Troubleshoot -> Diagnostics, upload the results

Then come back here, click new issue at the top
Fill in the template, include your diagnostic ID
Thanks :)

**Comment by cristianmad** [comment]
@mikeparker That is awesome!
Thanks, will do.

**Comment by FrankSzendzielarz** [comment]
Great. First time I heard of this tool. I guess I will upload every time it happens, as it is sporadic and can obviously be from multiple causes. I guess I will open a new issue for each upload, just in case they all turn out to be different causes to help you with your tracking.

On 3 Sep 2019 17:49, Cristian <notifications@github.com> wrote:

@mikeparker<https://github.com/mikeparker> That is awesome!
Thanks, will do.

**Comment by marcel-benthin** [comment]
I have this issue since monday (2019-09-09)
Diagnostics ID: 5A683DFF-B81A-4CBF-8401-57E4D2166E67/20190911074652
Setting memory from 2GB to 4GB didn’t work (with 2CPUs).
Reducing to 1 CPU (with 2GB memory) lowered tho overall CPU consumption but that one CPU is fully utilized, so no change. 
More swap or disk also didn’t help.

It started after upgrading to 2.1.0.2

**Comment by babeal** [comment]
Same issue:  vmmem takes 100% of a cpu causing docker start to timeout and throw exception 
Diagnostics ID: 35013B44-FAE4-4190-9CD6-BA048F32008B/20190920192605

Brandt new host image and 0 running containers.

**Comment by ffb255** [comment]
Same issue here.
Docker desktop consume CPU up to 80%!

**Comment by matomesc** [comment]
Having the same issue, 500k+ page faults / sec, restarting docker desktop temporarily fixes this.

**Comment by StingyJack** [comment]
> Furthermore, I am part of prioritization at docker and I can assure you, adding 'me too' comments into an old closed bug will not help us fix any issue

@mikeparker - thats because to you its an "old closed bug" but its not that to the person experiencing it right now that needs help because something you made is malfunctioning.  The solution to your perception problem (and one of the few things Zendesk does right) is to reopen the issue when someone says they are still experiencing it. For support and development teams this could seem like a total drag (and i can totally understand this viewpoint), but the goal is not to cater to their feelings, the goal is customer satisfaction. 

Also problems like this one are usually not due to anything the customer is doing. This is something wrong with the software that was produced, so keeping the same issue open makes far more sense because its an issue on Docker's end that has been present for a long time and is still not resolved.

**Comment by djereg** [comment]
@StingyJack You're absolutely right!

Same issue here!

Docker Desktop 2.2.0.5

**Comment by Patchoulino** [comment]
same issue, I disabled docker to auto start and next time I restarted my computer it was getting very hot without anything running, for some reason docker was running using 40% CPU and 100% Memory.

not sure why it was using resources when docker wasn't even enabled to auto start

Uninstalled fixed the issue...

**Comment by Suupa** [comment]
@mikeparker It's 30/05/2020 now and this problem still persists. I can't for the life of me understand why such a critical issue isn't reopened when it clearly hasn't been fixed. Please reopen this issue.

**Comment by klerpi** [comment]
This is still happening in the latest update. 

The thing is, before updating it never happened, but now Vmmem is consuming most of my memory and refuses to close after I closed Docker.

**Comment by aalbericio** [comment]
Just updated to latest 2020 update, WSL2, docker desktop stopped and vmmem up to 2Gb :(

**Comment by MGMehdi** [comment]
@aalbericio same

**Comment by LyonsDo** [comment]
Chews up the CPU and seems to take forever to uninstall.  Leaves a bad taste.

**Comment by MGMehdi** [comment]
Using a VM with Linux for docker use less ressources than docker for windows...

**Comment by LyonsDo** [comment]
Thanks for the reply. It's the use of resources when no container was
active that bothered me.

Anyway it's uninstalled now.

**Comment by Schneeple** [comment]
I recently upgraded to the Windows 2004 update and using the WSL2 and Just got this error today. 
There are no running containers. The CPU and Memory were both at 100% but the CPU has dropped just leaving the Memory at 100%.

![image](https://user-images.githubusercontent.com/41080582/85213101-cee13080-b30e-11ea-93dd-ea776ffa7786.png)

**Comment by geoffreyvanwyk** [comment]
Should be reopenend.

**Comment by Animeshz** [comment]
Same issue, on my 8GB machine approx 2GB has been eaten by Docker and its Vmmem. Running on Windows 10 Home 2004 with WSL2 (Ubuntu 18.04).
![Screenshot (5)](https://user-images.githubusercontent.com/26714676/85977337-0bd9b100-b9fa-11ea-8a9c-d73b4114a854.png)

**Comment by mariocalin** [comment]
3.5 GB of memory usage with mysql and mongo instances with really little use.

**Comment by FaysalM** [comment]
52 GB of memory. This is insane. No dockers running, I do have a few visual studio open where I build dockers. However last build was 1 hour ago and the memory is still stuck in something.
![image](https://user-images.githubusercontent.com/6191014/87043650-ba6ec480-c238-11ea-9a28-15441850ddc9.png)

**Comment by jwhipp** [comment]
@mariocalin @FaysalM If you guy are using **WSL 2** as your back end, see https://github.com/microsoft/WSL/issues/4166 , WSL 2 memory consumption isn't a docker issue. 

You can edit your ~\.wslconfig to limit ram, it's a known issue with WSL 2 right now.

**Comment by mariocalin** [comment]
> 
> 
> @mariocalin @FaysalM If you guy are using **WSL 2** as your back end, see [microsoft/WSL#4166](https://github.com/microsoft/WSL/issues/4166) , WSL 2 memory consumption isn't a docker issue.
> 
> You can edit your ~.wslconfig to limit ram, it's a known issue with WSL 2 right now.

You are absolutely right. I edited the wslconfig and now it is controlled. I guess I was expecting some help in docker-desktop app.

**Comment by jshin313** [comment]
> @mariocalin @FaysalM If you guy are using **WSL 2** as your back end, see [microsoft/WSL#4166](https://github.com/microsoft/WSL/issues/4166) , WSL 2 memory consumption isn't a docker issue.
> 
> You can edit your ~.wslconfig to limit ram, it's a known issue with WSL 2 right now.

Thanks so much! This helped tremendously.

**Comment by djereg** [comment]
Until the last 2.3.0.4 update everything was O.K., but now Vmmem is consuming ~20% CPU constantly without any running containers.

Diagnostic ID: D1DAAED1-1AC1-44AD-B43F-D4ED4FADC551/20200729084411

**Comment by srepollock** [comment]
I am also seeing issues with this. I don't know why this issue is closed, it seems very active.

Running VSCode with the integrated terminal running WSL2, every command hangs. Any time I get to run a command, the CPU usage shoots up to 60-80% on my machine, and the command has already finished printing. It just hangs empty. I get up to ~2GB RAM usage. I'm setting it in my `~.wslconfig` to be 4GB max ram, but it doesn't seem to affect things. 


I've cleared/purged the Docker data for everything (WSL, Windows, etc), and followed the `.wslconfig` to no avail.

I'm hoping to get VSCode working with integrated container development, but at this time, it's not feasible given how useless my system is while things are just waiting.

This is still an issue.

### Versions
WSL: 2
Windows10Pro: Version 2004 (OS Build 19041.421)
Docker for Windows: 2.3.0.4 (46911)

**Diagnostic ID: B24E1CE7-244E-4D89-A5C9-BB310292C5BD/20200731170410**

**Comment by topotru** [comment]
> Until the last 2.3.0.4 update everything was O.K., but now Vmmem is consuming ~20% CPU constantly without any running containers.
> 
> Diagnostic ID: D1DAAED1-1AC1-44AD-B43F-D4ED4FADC551/20200729084411

I have the same issue.
Until the last 2.3.0.4 update everything was O.K., but now Vmmem is consuming ~30% CPU constantly without any running containers.

Win10pro 1909
Hyper-V (no wsl2)

Diagnostic ID: 5D8D09EC-5C03-4897-B6FF-C2FC58624C44/20200804125505
